proxyd/consensus_tracker.go

```diff
@@ -21,101 +26,311 @@ type ConsensusTracker interface {
+			if err != nil && err != redis.Nil {
+				log.Error("failed to read the remote state", "err", err)
+				return
+			}
+			if val == "" {
+				log.Error("remote state is missing (recent leader election maybe?)")
+				return
+			}
+			state := &ConsensusTrackerState{}
+			err = json.Unmarshal([]byte(val), state)
+			if err != nil {
+				log.Error("failed to unmarshal the remote state", "err", err)
+				return
+			}
+			ct.remote.mutex.Lock()
+			defer ct.remote.mutex.Unlock()
+			ct.remote.state.update(state)
+			log.Debug("updated state from remote", "state", val, "leader", leaderName)
+		}
+	} else {
+		if !ct.local.Valid() {
+			log.Warn("local state is not valid or behind remote, skipping")
+			return
+		}
+		if ct.remote.Valid() && ct.local.Behind(ct.remote) {
+			log.Warn("local state is behind remote, skipping")
+			return
+		}
+
+		log.Info("lock not found, creating a new one")
+
+		mutex := rs.NewMutex(key,
+			redsync.WithExpiry(ct.lockPeriod),
+			redsync.WithFailFast(true),
+			redsync.WithTries(1))
+
+		// nosemgrep: missing-unlock-before-return
+		// this lock is hold indefinitely, and it is extended until the leader dies
+		if err := mutex.Lock(); err != nil {
+			log.Debug("failed to obtain lock", "err", err)
+			ct.leader = false
+			return
+		}
+
+		log.Info("lock acquired", "mutex", mutex.Name(), "val", mutex.Value())
+		ct.redlock = mutex
+		ct.leader = true
+		ct.postPayload(mutex.Value())
 	}
 }
 
 func (ct *RedisConsensusTracker) key(tag string) string {
-	return fmt.Sprintf("consensus:%s:%s", ct.backendGroup, tag)
+	return fmt.Sprintf("consensus:%s:%s", ct.namespace, tag)
 }
 
 func (ct *RedisConsensusTracker) GetLatestBlockNumber() hexutil.Uint64 {
-	return hexutil.Uint64(hexutil.MustDecodeUint64(ct.client.Get(ct.ctx, ct.key("latest")).Val()))
+	return ct.remote.GetLatestBlockNumber()
 }
 
 func (ct *RedisConsensusTracker) SetLatestBlockNumber(blockNumber hexutil.Uint64) {
-	ct.client.Set(ct.ctx, ct.key("latest"), blockNumber, 0)
+	ct.local.SetLatestBlockNumber(blockNumber)
 }
 
 func (ct *RedisConsensusTracker) GetSafeBlockNumber() hexutil.Uint64 {
-	return hexutil.Uint64(hexutil.MustDecodeUint64(ct.client.Get(ct.ctx, ct.key("safe")).Val()))
+	return ct.remote.GetSafeBlockNumber()
 }
 
 func (ct *RedisConsensusTracker) SetSafeBlockNumber(blockNumber hexutil.Uint64) {
-	ct.client.Set(ct.ctx, ct.key("safe"), blockNumber, 0)
+	ct.local.SetSafeBlockNumber(blockNumber)
 }
 
 func (ct *RedisConsensusTracker) GetFinalizedBlockNumber() hexutil.Uint64 {
-	return hexutil.Uint64(hexutil.MustDecodeUint64(ct.client.Get(ct.ctx, ct.key("finalized")).Val()))
+	return ct.remote.GetFinalizedBlockNumber()
 }
 
 func (ct *RedisConsensusTracker) SetFinalizedBlockNumber(blockNumber hexutil.Uint64) {
-	ct.client.Set(ct.ctx, ct.key("finalized"), blockNumber, 0)
+	ct.local.SetFinalizedBlockNumber(blockNumber)
+}
+
+func (ct *RedisConsensusTracker) postPayload(mutexVal string) {
+	ct.remote.mutex.Lock()
```

Comment:
Ditto here - can we move this mutex management inside the `remote` object?

Reply:
Moved as well

Reply:
I think you may have forgotten this one - since you moved the mutex inside `update()` I'm pretty sure this will deadlock.

Reply:
Nice catch